### PR TITLE
clp-s: Add support for decompression in ascending timestamp order.

### PR DESCRIPTION
**Pull request by gibber9809**
# Description
This PR adds support to decompress a clp-s archive in timestamp order with the `--ordered` command line option. This is implemented by loading all tables into memory, putting them into a min-heap ordered by the timestamp of the next record in each table, and popping from the heap until all records have been decompressed. The timestamp is pulled from the authoritative timestamp column specified at compression time, and if a table does not contain such a column the timestamp is 0 for ordering purposes.

Note: we don't sort by timestamp within each table, so the logs can end up slightly out of timestamp order (this is probably preferred anyway, since it brings us closer to log order).

This seems to add minimal decompression speed overhead (about 2.5%) for fairly typical archives, though would likely have higher overhead for large archives or archives with many tables (due to extra memory allocations compared to decompressing one table at a time). Similarly there can be significant memory overhead, particularly for large archives. 

# Validation performed
- Validated that records end up sorted in timestamp order after decompression
- Validated that in order decompression results in identical set of records compared to out of order decompression

### CODE REVIEW

File: components/core/src/clp_s/ArchiveReader.cpp [hunk 1]
```diff
@@ -92,14 +92,39 @@ SchemaReader& ArchiveReader::read_table(
...
+    for (auto& [id, table_metadata] : m_id_to_table_metadata) {
+        rit->reset(new SchemaReader());
+        create_schema_reader(**rit, id, true, true);
+
+        m_tables_file_reader.try_seek_from_begin(table_metadata.offset);
+        m_tables_decompressor.open(m_tables_file_reader, cDecompressorFileReadBufferCapacity);
+        (*rit)->load(m_tables_decompressor, table_metadata.uncompressed_size);
+        m_tables_decompressor.close_for_reuse();
+        ++rit;
+    }
```
Comment (wraymo) [inline components/core/src/clp_s/ArchiveReader.cpp:126]:
```suggestion
    std::vector<std::shared_ptr<SchemaReader>> readers;
    readers.reserve(m_id_to_table_metadata.size());
    for (const auto& [id, table_metadata] : m_id_to_table_metadata) {
        auto schema_reader = std::make_shared<SchemaReader>();
        create_schema_reader(*schema_reader, id, true, true);

        m_tables_file_reader.try_seek_from_begin(table_metadata.offset);
        m_tables_decompressor.open(m_tables_file_reader, cDecompressorFileReadBufferCapacity);
        schema_reader->load(m_tables_decompressor, table_metadata.uncompressed_size);
        m_tables_decompressor.close_for_reuse();

        readers.push_back(std::move(schema_reader));
    }
```
Personally, I would write the code block in this way.. If you want to keep it, do you mind change line 1219 to `*rit = std::make_shared<SchemaReader>();`?

File: components/core/src/clp_s/JsonConstructor.hpp [hunk 2]
```diff
@@ -49,8 +51,11 @@ class JsonConstructor {
     void store();
 
 private:
+    void construct_in_order(FileWriter& writer);
```
Comment (wraymo) [inline components/core/src/clp_s/JsonConstructor.hpp:54]: Can you add a description for this method?

File: components/core/src/clp_s/ArchiveReader.cpp (hunk 1 above)
Comment (gibber9809) [reply components/core/src/clp_s/ArchiveReader.cpp:126]: I like your way better too.

File: components/core/src/clp_s/CommandLineArguments.cpp [hunk 3]
```diff
@@ -286,6 +286,14 @@ CommandLineArguments::parse_arguments(int argc, char const** argv) {
             );
             extraction_options.add(input_options);
 
+            po::options_description decompression_options("Decompression Options");
+            decompression_options.add_options()(
+                    "ordered",
+                    po::bool_switch(&m_ordered_decompression),
+                    "Enable in-order decompression for this archive"
```
[the remark this answers is not in the document]
Comment (gibber9809) [reply components/core/src/clp_s/CommandLineArguments.cpp:293]: Yeah, I had some vague idea that we'd transparently use timestamp or MOT depending on what was available in the archive, but being explicit is best for now.

**Review by wraymo — COMMENTED** [review]
Great work! Since we only support ascending order for timestamps, should we explicitly state this in the command-line description and the commit message?

**Review by wraymo — APPROVED** [review]
For the commit message, what about "clp-s: Add support for decompression in ascending timestamp order."?

**Pull request merged**